File: 4/swift/main.swift
import Foundation

// MARK: - Models

final class BingoBoard {
  let rows: [[Int]]

  var calledNumbers: Set<Int> = []

  func call(number: Int) {
    if rows.contains(where: { $0.contains(number) }) {
      calledNumbers.insert(number)
    }
  }

  init(lines: [String]) {
    rows = lines.map { line in
      line
        .components(separatedBy: .whitespaces)
        .compactMap(Int.init)
    }
  }

  var columns: [[Int]] {
    return (0..<5).map { column in rows.map { $0[column] } }
  }

  var didWin: Bool {
    return rows.contains(where: { calledNumbers.isSuperset(of: $0) }) ||
      columns.contains(where: { calledNumbers.isSuperset(of: $0) })
  }

  var unmarkedSum: Int {
      Set(rows.flatMap { $0 })
      .subtracting(calledNumbers)
      .reduce(0, +)
  }
}

// MARK: - Get Scores

func getScores(boards: [BingoBoard], calledNumbers: [Int]) -> (first: Int, last: Int) {
  var firstScore: Int?
  var winningBoards = [Int]()
  for number in calledNumbers {
    for (boardIndex, board) in boards.enumerated() where !winningBoards.contains(boardIndex) {
      board.call(number: number)
      if board.didWin {
        winningBoards.append(boardIndex)
        if firstScore == nil {
          firstScore = board.unmarkedSum * number
        } else if winningBoards.count == boards.count,
          let firstScore = firstScore
        {
          return (first: firstScore, last: board.unmarkedSum * number)
        }
      }
    }
  }

  fatalError("Not all boards won")
}

// MARK: - Read Input

let contents = try String(contentsOfFile: "input.txt", encoding: .utf8)
let lines = contents.components(separatedBy: .newlines)

// MARK: - Parse Called Numbers

let calledNumbers = lines[0]
  .split(separator: ",")
  .compactMap { Int($0) }

// MARK: - Parse Bingo Boards

let boards = stride(from: 2, to: lines.count, by: 6)
  .map { startLine -> [String] in
    [
      lines[startLine],
      lines[startLine + 1],
      lines[startLine + 2],
      lines[startLine + 3],
      lines[startLine + 4],
    ]
  }
  .map(BingoBoard.init)

// MARK: - Parts 1 & 2

let (first, last) = getScores(boards: boards, calledNumbers: calledNumbers)
print("first score: \(first)")
print("last score: \(last)")

File: 4/cpp/main.cc
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
using namespace std;

// MARK: - Models

struct BingoBoard {
  vector<vector<int>> rows;
  vector<int> called_numbers;

  void call(int number) {
    called_numbers.push_back(number);
  }

  vector<vector<int>> columns() {
    vector<vector<int>> columns;
    for (int column_index = 0; column_index < 5; ++column_index)
    {
      vector<int> column;
      for (auto i = rows.begin(); i != rows.end(); ++i)
      {
        vector<int> row = *i;
        column.push_back(row[column_index]);
      }
      columns.push_back(column);
    }
    return columns;
  }

  bool did_win() {
    for (auto i = rows.begin(); i != rows.end(); ++i)
    {
      int matches = 0;
      for (auto n = called_numbers.begin(); n != called_numbers.end(); ++n)
      {
        if (count(i->begin(), i->end(), *n) > 0)
          matches++;
      }
      if (matches == 5)
        return true;
    }

    vector<vector<int>> cols = columns();
    for (auto i = cols.begin(); i != cols.end(); ++i)
    {
      int matches = 0;
      for (auto n = called_numbers.begin(); n != called_numbers.end(); ++n)
      {
        if (count(i->begin(), i->end(), *n) > 0)
          matches++;
      }
      if (matches == 5)
        return true;
    }

    return false;
  }

  int unmarked_sum() {
    int sum = 0;

    for (auto r = rows.begin(); r != rows.end(); ++r)
    {
      for (auto c = r->begin(); c != r->end(); ++c)
      {
        if (count(called_numbers.begin(), called_numbers.end(), *c) == 0)
          sum += *c;
      }
    }

    return sum;
  }
};

// MARK: - Get Scores

pair<int, int> get_scores(vector<BingoBoard>& boards, vector<int> called_numbers) {
  int first = 0;
  vector<int> winning_boards;
  for (auto n = called_numbers.begin(); n != called_numbers.end(); ++n)
  {
    for (int board_index = 0; board_index < boards.size(); ++board_index)
    {
      if (count(winning_boards.begin(), winning_boards.end(), board_index) > 0)
      {
        continue;
      }

      BingoBoard& board = boards[board_index];
      board.call(*n);
      if (board.did_win()) {
        winning_boards.push_back(board_index);
        if (winning_boards.size() == 1) {
          first = *n * board.unmarked_sum();
        } else if (winning_boards.size() == boards.size()) {
          return {first, *n * board.unmarked_sum()};
        }
      }
    }
  }

  cout << "Not all boards won" << endl;
  abort();
}

// MARK: - Split

vector<string> split(string str, string token){
    vector<string>result;
    while(str.size()){
        int index = str.find(token);
        if (index != string::npos) {
            result.push_back(str.substr(0,index));
            str = str.substr(index+token.size());
            if (str.empty()) result.push_back(str);
        } else {
            result.push_back(str);
            str = "";
        }
    }
    return result;
}

int main() {
  // MARK: - Read Input

  ifstream is("input.txt");
  string str;
  vector<BingoBoard> boards;
  vector<int> called_numbers;
  vector<string> next_board_lines;
  while(getline(is, str)) {
    // MARK: - Parse Called Numbers
    if (called_numbers.empty()) {
      vector<string> called_numbers_str = split(str, ",");
      for (auto i = called_numbers_str.begin(); i != called_numbers_str.end(); ++i)
      {
        called_numbers.push_back(stoi(*i));
      }
      continue;
    } else if (str.empty()) {
      continue;
    }

    // MARK: - Parse Bingo Boards

    next_board_lines.push_back(str);
    if (next_board_lines.size() == 5) {
      BingoBoard board;
      for (auto i = next_board_lines.begin(); i != next_board_lines.end(); ++i)
      {
        stringstream stream(*i);
        int number;
        vector<int> numbers;
        while (stream >> number) {
          numbers.push_back(number);
        }
        board.rows.push_back(numbers);
      }
      
      boards.push_back(board);
      next_board_lines.clear();
    }
  }

  // MARK: - Parts 1 & 2

  pair<int, int> scores = get_scores(boards, called_numbers);
  int first = scores.first;
  cout << "first score: " << first << endl;
  int last = scores.second;
  cout << "last score: " << last << endl;
  return 0;
}
